### ISSUE TITLE
try/catch inconsistent

### ISSUE THREAD

**Issue by strux-net** [vim/vim#3142]
to demonstrate it I have a file with the following content
```
try
    sleep 2
  catch /^Vim:Interrupt$/
    echo 'interupted'
  finally
    echo 'continuing'
endtry
" I wonder why I do not see "Press enter" if i press ^C within 2 seconds (to interrupt the sleep)
call input('Press enter')
call input('press enter again')

```
`vim -u NONE -U NONE -c "so file.vim"` rsp.
`gvim -u NONE -U NONE -c "so file.vim"`

Observation : gvim shows the "Press enter" line, while vim does not.

If it matters, this is :

```
VIM - Vi IMproved 8.1 (2018 May 18, compiled Jul  1 2018 04:17:01)
Included patches: 1-127, 129-133
```

**Comment by brammool** [comment]
> to demonstrate it I have a file with the following content
> ```
> try
>     sleep 2
>   catch /^Vim:Interrupt$/
>     echo 'interupted'
>   finally
>     echo 'continuing'
> endtry
> " I wonder why I do not see "Press enter" if i press ^C within 2 seconds (to interrupt the sleep)
> call input('Press enter')
> call input('press enter again')
> 
> ```
> `vim -u NONE -U NONE -c "so file.vim"` rsp.
> `gvim -u NONE -U NONE -c "so file.vim"`
> 
> Observation : gvim shows the "Press enter" line, while vim does not.
> 
> If it matters, this is :
> 
> ```
> VIM - Vi IMproved 8.1 (2018 May 18, compiled Jul  1 2018 04:17:01)
> Included patches: 1-127, 129-133
> ```

Most likely the CTRL-C causes script execution to be aborted.
When try/catch was added this was kept backwards compatible.
Thus after the catch and finally statements execution of the script
stops.

You can try putting this in a function, then perhaps only the function
aborts.

-- 
hundred-and-one symptoms of being an internet addict:
173. You keep tracking down the email addresses of all your friends
     (even childhood friends).

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by strux-net** [comment]
I don't think that it is related to `source`.
The script is not aborted - the second `input()` gets executed.
Only the first `input('press enter')` does not show up.
Also, wrapping it in a function doesn't change behaviour
```

function Foo()
  try
      sleep 2
    catch /^Vim:Interrupt$/
      echo 'interupted'
    finally
      echo 'continuing'
  endtry
  " I wonder why I do not see "Press enter" if i press ^C within 2 seconds (to interrupt the sleep)
  call input('Press enter')
  call input('press enter again')
endfunction
```


`call Foo()`

**Comment by brammool** [comment]
Michael Naumann wrote:

> I don't think that it is related to `source`.
> The script is not aborted - the second `input()` gets executed.
> Only the first `input('press enter')` does not show up.
> Also, wrapping it in a function doesn't change behaviour
> ```
> 
> function Foo()
>   try
>       sleep 2
>     catch /^Vim:Interrupt$/
>       echo 'interupted'
>     finally
>       echo 'continuing'
>   endtry
>   " I wonder why I do not see "Press enter" if i press ^C within 2 seconds (to interrupt the sleep)
>   call input('Press enter')
>   call input('press enter again')
> endfunction
> ```
> 
> 
> `call Foo()`

OK, I see it now (I thought you said was was working OK in the GUI and
not working in the terminal).

Looks like in this specific case we never read the CTRL-C from the input
stream before calling input().

-- 
Some of the well known MS-Windows errors:
	EHUH		Unexpected error
	EUSER		User error, not our fault!
	EGOD		Horrible problem, god knows what has happened
	EERR		Errornous error: nothing wrong

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by strux-net** [comment]
yes, sorry, I should have phrased more clearely.

**Comment by brammool** [comment]
Fixed with patch 8.1.0158
